syntax = "proto3";

import "protos/Article/Article.proto";

message GetArticlesRequest {
    string client_uuid = 1;

    // article info:
    string type = 2;
    string author = 3;
    Date date = 4;   
}

message GetArticlesResponse {
    repeated Article article_list = 1;
}

message PublishArticleRequest {
    string client_uuid = 1;
    Article article = 2;
}

message PublishArticleResponse {
    enum Status {
        FAILED = 0;
        SUCCESS = 1;
    }

    Status status = 1;
}

// JoinServer Messages;

message ServerJoinRequest {
    string client_uuid = 1;
    bool is_server = 2;
}

message ServerJoinResponse {
    enum Status {
        FAILED = 0;
        SUCCESS = 1;
    }
    
    Status status = 1;
}

// LeaveServer Messages
message ServerLeaveRequest {
    string client_uuid = 1;
}

message ServerLeaveResponse {
    enum Status {
        FAILED = 0;
        SUCCESS = 1;
    }
    
    Status status = 1;
}

service ClientServer {
    rpc GetArticles (GetArticlesRequest) returns (GetArticlesResponse);
    rpc PublishArticle (PublishArticleRequest) returns (PublishArticleResponse);

    // Manvi RPCs
    rpc JoinServer (ServerJoinRequest) returns (ServerJoinResponse);
    rpc LeaveServer (ServerLeaveRequest) returns (ServerLeaveResponse);
}
